### ISSUE TITLE
NPM5 Permissions Issue

### ISSUE THREAD

**Issue by xjamundx** [npm/npm#16896]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

With npm 5.0.0 when I try to install a handful of packages they give different permissions than when I installed with NPM2/3/4.

#### How can the CLI team reproduce the problem?

With `NPM 5.0.1`
```
npm i meddleware
ls -lsa node_modules/meddleware/index.js
16 -rw-------  1 me  110163442  7569 Aug  7  2015 node_modules/meddleware/index.js
```
With `NPM 4.6.1`
```
npm i meddleware
ls -lsa node_modules/meddleware/index.js
16 -rw-r--r--  1 me  110163442  7569 Aug  7  2015 node_modules/meddleware/index.js
```
### Random Notes

This might be a feature. It's also something that can be addressed on a per-package basis. Currently I've filed bugs on repos that have this issue. But I thought I'd mention it here as well.

Here's another example of where this is happening in the wild (also seemed to be an issue with `yarn`):
https://github.com/jonschlinkert/array-unique/issues/5

I'm not sure if it's related to #16766 or not.

### supporting information:

 - `npm -v` prints:
5.0.1

 - `node -v` prints:
6.10.3

 - `npm config get registry` prints:
`https://registry.npmjs.org/`

 - Windows, OS X/macOS, or Linux?:
mac

 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [x] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by zkat** [comment]
This is probably an issue with [pacote not getting permissions right during extraction](https://github.com/zkat/pacote/blob/latest/lib/extract-stream.js). This might also have to do with `tar-fs` itself, which both Yarn and npm use for extracting tarballs -- it might just be a matter of configuring this right to emulate npm's old behavior (which I or someone else adventurous would have to look up: it's [somewhere in `lib/utils/tar.js`](https://github.com/npm/npm/blob/v4.6.1/lib/utils/tar.js#L230-L437) with the whole `fMode` and `dMode` thing.

Then you "just" need to [pass the options through to `pacote`](https://github.com/npm/npm/blob/latest/lib/config/pacote.js) (which isn't being done right now), and it should use it, since it's already passing the mode information to `tar-fs`.

This looks like a nice patch for someone to PR. Please let me know if there's any questions about how to find something required for this.

**Comment by max-mapper** [comment]
for funsies I compared what you get by double clicking the `meddleware` tarball in mac os 10.10:

```
🐈  ls -alh package
total 64
drwxr-xr-x@   8 max  staff   272B Jun  1 16:22 .
drwxr-xr-x+ 214 max  staff   7.1K Jun  1 16:26 ..
-rw-r--r--@   1 max  staff   1.7K Oct 20  2014 LICENSE.txt
-rw-------@   1 max  staff    10K Aug  7  2015 README.md
-rw-r--r--@   1 max  staff   598B Oct 20  2014 SECURITY.md
-rw-------@   1 max  staff   7.4K Aug  7  2015 index.js
drwxr-xr-x@   4 max  staff   136B Jun  1 16:22 lib
-rw-------@   1 max  staff   1.3K Aug  7  2015 package.json
```

versus what you get with this code:

```js
var fs = require('fs')
var tar = require('tar-fs')
process.stdin.pipe(tar.extract('./package2'))
```

```
🐈  ls -alh package2/package/
total 64
drwxr-xr-x  8 max  staff   272B Jun  1 16:26 .
drwxr-xr-x  3 max  staff   102B Jun  1 16:26 ..
-rw-r--r--  1 max  staff   1.7K Oct 20  2014 LICENSE.txt
-rw-------  1 max  staff    10K Aug  7  2015 README.md
-rw-r--r--  1 max  staff   598B Oct 20  2014 SECURITY.md
-rw-------  1 max  staff   7.4K Aug  7  2015 index.js
drwxr-xr-x  4 max  staff   136B Jun  1 16:26 lib
-rw-------  1 max  staff   1.3K Aug  7  2015 package.json
```

Looks like they match which suggests tar-fs is doing the right thing

**Comment by xjamundx** [comment]
@maxogden I think previous versions of NPM normalized the permissions in some way. I wasn't sure if this feature was purposefully removed or not!

**Comment by zkat** [comment]
@xjamundx that is correct: the code linked above seems to do that normalization. I guess I can ask @isaacs why that normalization was added, since he's the most familiar with `tar.js` here.

**Comment by chrmoritz** [comment]
I wanted to just report the same issue myself, because this can actual cause package installation to fail, which we encountered when upgrading npm to version 5 in homebrew.

Refs: https://github.com/Homebrew/homebrew-core/pull/14152 and https://github.com/Homebrew/homebrew-core/pull/14085#issuecomment-305574029
 
npm@5 will throw (on unix) if you try to install a package (such as `easy-table@0.0.1`) which contains incorrect file system permissions (in the `.tgz` tarball downloaded from the registry), while npm@4 succeeded ignoring the incorect permissions.

Reproducible by running `npm install easy-table@0.0.1` with npm@5 as a non-root user (you won't have +x permission for the lib directory)
 ```
  > npm install easy-table@0.0.1
npm WARN enoent ENOENT: no such file or directory, open '/Users/chris/Desktop/test/package.json'
npm WARN test No description
npm WARN test No repository field.
npm WARN test No README data
npm WARN test No license field.

npm ERR! path /Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall open
npm ERR! Error: EACCES: permission denied, open '/Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js'
npm ERR!  { Error: EACCES: permission denied, open '/Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js'
npm ERR!   cause: 
npm ERR!    { Error: EACCES: permission denied, open '/Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js'
npm ERR!      errno: -13,
npm ERR!      code: 'EACCES',
npm ERR!      syscall: 'open',
npm ERR!      path: '/Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js' },
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/Users/chris/Desktop/test/node_modules/.staging/easy-table-3711f69a/lib/table.js' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/chris/.npm/_logs/2017-06-01T20_33_00_832Z-debug.log
 ```

```
> wget https://registry.npmjs.org/easy-table/-/easy-table-0.0.1.tgz
> tar xfz easy-table-0.0.1.tgz 
> ls -la package/
total 24
drwxr-xr-x  7 chris  staff   238  2 Jun 11:22 .
drwxr-xr-x  4 chris  staff   136  2 Jun 11:22 ..
-rw-r--r--  1 chris  staff  1646 13 Mär  2012 README.md
-rw-r--r--  1 chris  staff   510 13 Mär  2012 example.js
drw-r--r--  3 chris  staff   102 13 Mär  2012 lib
-rw-r--r--  1 chris  staff   385 13 Mär  2012 package.json
drw-r--r--  3 chris  staff   102 13 Mär  2012 test
```
 
This has broken our `azure-cli` formula and was fixed by patching it to temporarily use the github source tarball instead of the npm registry tarball for `easy-table` in https://github.com/Homebrew/homebrew-core/pull/14152.

* npm debug log for easy-table@0.0.1: https://gist.github.com/chrmoritz/b21612a74f9d278f0fa14b243a5b0fe9
* npm debug log for azure-cli: https://gist.github.com/chrmoritz/a01420db1b21f347011f0f0b96d0d7f8

**Comment by marklagendijk** [comment]
I have a [similar issue](https://github.com/Medium/phantomjs/issues/707) when installing phantomjs-prebuilt globally, using sudo.

```
sudo npm install -g phantomjs-prebuilt
/usr/bin/phantomjs -> /usr/lib/node_modules/phantomjs-prebuilt/bin/phantomjs

> phantomjs-prebuilt@2.1.14 install /usr/lib/node_modules/phantomjs-prebuilt
> node install.js

Considering PhantomJS found at /usr/bin/phantomjs
Looks like an `npm install -g`
Could not link global install, skipping...
Download already available at /tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2
Verified checksum of previously downloaded file
Extracting tar contents (via spawned process)
Removing /usr/lib/node_modules/phantomjs-prebuilt/lib/phantom
Copying extracted folder /tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2-extract-1496325965675/phantomjs-2.1.1-linux-x86_64 -> /usr/lib/node_modules/phantomjs-prebuilt/lib/phantom
Phantom installation failed { Error: EACCES: permission denied, link '/tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2-extract-1496325965675/phantomjs-2.1.1-linux-x86_64' -> '/usr/lib/node_modules/phantomjs-prebuilt/lib/phantom'
    at Error (native)
  errno: -13,
  code: 'EACCES',
  syscall: 'link',
  path: '/tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2-extract-1496325965675/phantomjs-2.1.1-linux-x86_64',
  dest: '/usr/lib/node_modules/phantomjs-prebuilt/lib/phantom' } Error: EACCES: permission denied, link '/tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2-extract-1496325965675/phantomjs-2.1.1-linux-x86_64' -> '/usr/lib/node_modules/phantomjs-prebuilt/lib/phantom'
    at Error (native)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! phantomjs-prebuilt@2.1.14 install: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the phantomjs-prebuilt@2.1.14 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```
I am using:
- OS: Ubuntu 16.04
- Node: v6.10.3, installed via NodeSource
- npm: 5.0.1
- `sudo npm install -g phantomjs-prebuilt`, with default NodeSource installation, `sudo` is needed for installing global packages.

**Comment by chrmoritz** [comment]
@marklagendijk Looks like your issue is unrelated to the issue reported here, because its caused by the [phantomjs-prebuilt's install.js script](https://github.com/Medium/phantomjs/blob/master/install.js#L289) which extracts the tarball by spawning tar in a child process. This issue here is more about npm (or more specific pacote) not normalizing permission anymore, when extracting tarballs.

Your issue is likely caused by `/usr/lib/node_modules/phantomjs-prebuilt/lib` being owned by root (because you installed with sudo) and phantomjs-prebuilt's install.js script trying to link its downloaded prebuild as a non-root user to it.

**Comment by kanongil** [comment]
The missing normalization is causing breakage on my setup as well, and I have had to revert to v4 for the time being.

This issue is especially critical when the npm install has `root` permissions, where it will set the user/group ids as well.

**Comment by dhniels** [comment]
Looks like im having the same issue. Rxjs installs with -rw-r-----  https://github.com/npm/npm/issues/17071#issuecomment-306963557

**Comment by zkat** [comment]
weeeeelp looks like we should bring that back, I guess ;)

**Comment by zkat** [comment]
Hey y'all. I've pushed out a new npm canary at `npmc@5.0.3-canary.11`. This _should_ be setting permissions the way npm used to, and will likely be released with the next version of npm (tomorrow?) Would any of y'all be able to verify that the fix works for you? (you can install this with `npm i -g npmc@latest` and run it with `npmc install ...`. It won't affect your existing npm)

**Comment by chrmoritz** [comment]
I can confirm that `npmc install easy-table@0.0.1` and `npmc install -g azure-cli` is working fine again (without permission issues) with the latest `npmc`. Thanks for fixing it so quickly.

**Comment by zkat** [comment]
woo. Thanks for testing it out!

**Comment by nickwalkmsft** [comment]
I don't know to what extent https://github.com/npm/npm/issues/16892 might be related - a number of issues seem to center around permissions - but I gave `npmc` a shot and it still repros.

Running as root in the node:8 container with Docker's userns remapping enabled I was able to successfully do `npm i -g npmc@latest`, but both `npmc install -g azure-cli` (attempted since @chrmoritz indicated success) and `npmc install express` in an empty dir failed, both with some variant of `EINVAL: invalid argument, chown '.../node_modules/.staging/<some_dependency>/...'`

**Comment by xjamundx** [comment]
Testing...that solves it for our needs, thanks!

**Comment by cebor** [comment]
Does this also fix #16766 ?

**Comment by danielbeardsley** [comment]
The `npmc@latest` didn't solve the issue for phantomjs, though I still think it's related as it seems to work fine performing the same install using npm4

**Comment by olalonde** [comment]
Might this be related (npm@5 inside docker)? https://discuss.circleci.com/t/getting-a-weird-einval-error-with-npm-on-circleci-2-0/13465

```
...
npm WARN enoent SKIPPING OPTIONAL DEPENDENCY: ENOENT: no such file or directory, utime '/ci/project/node_modules/.staging/acorn-7a8d946c/dist/acorn.es.js'

npm ERR! path /ci/project/node_modules/.staging/winston-81b2de90
npm ERR! code EINVAL
npm ERR! errno -22
npm ERR! syscall chown
npm ERR! EINVAL: invalid argument, chown '/ci/project/node_modules/.staging/winston-81b2de90'

npm ERR! A complete log of this run can be found in:
npm ERR! /root/.npm/_logs/2017-06-16T17_58_08_634Z-debug.log
Exited with code 234
```

**Comment by zkat** [comment]
Closing this, since it's fixed.

**Comment by callumlocke** [comment]
@zkat in which npm version is it fixed? Do I have to install a beta version or something?

**Comment by zkat** [comment]
@callumlocke npmc, and it will be released in tomorrow's `5.1.0`
